File: rewrite-go/test/recipe_test.go
/*
 * Copyright 2025 the original author or authors.
 *
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://docs.moderne.io/licensing/moderne-source-available-license
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package test

import (
	"testing"
	"time"

	"github.com/stretchr/testify/require"

	"github.com/stretchr/testify/assert"

	"github.com/openrewrite/rewrite/rewrite-go/pkg/parser"
	"github.com/openrewrite/rewrite/rewrite-go/pkg/printer"
	"github.com/openrewrite/rewrite/rewrite-go/pkg/recipe"
	"github.com/openrewrite/rewrite/rewrite-go/pkg/test"
	"github.com/openrewrite/rewrite/rewrite-go/pkg/tree/java"
	"github.com/openrewrite/rewrite/rewrite-go/pkg/visitor"
)

type findFoo struct {
	recipe.Base
}

func (r *findFoo) Name() string        { return "org.openrewrite.golang.test.FindFoo" }
func (r *findFoo) DisplayName() string { return "Find foo identifiers" }
func (r *findFoo) Description() string {
	return "Marks all identifiers named `foo` with a search result."
}

func (r *findFoo) Editor() recipe.TreeVisitor {
	return visitor.Init(&findFooVisitor{})
}

type findFooVisitor struct {
	visitor.GoVisitor
}

func (v *findFooVisitor) VisitIdentifier(ident *java.Identifier, p any) java.J {
	ident = v.GoVisitor.VisitIdentifier(ident, p).(*java.Identifier)
	if ident.Name == "foo" {
		ident = ident.WithMarkers(java.FoundSearchResult(ident.Markers, "found foo"))
	}
	return ident
}

type renameFooToBar struct {
	recipe.Base
}

func (r *renameFooToBar) Name() string        { return "org.openrewrite.golang.test.RenameFooToBar" }
func (r *renameFooToBar) DisplayName() string { return "Rename foo to bar" }
func (r *renameFooToBar) Description() string { return "Renames all identifiers named `foo` to `bar`." }

func (r *renameFooToBar) Editor() recipe.TreeVisitor {
	return visitor.Init(&renameFooToBarVisitor{})
}

type renameFooToBarVisitor struct {
	visitor.GoVisitor
}

func (v *renameFooToBarVisitor) VisitIdentifier(ident *java.Identifier, p any) java.J {
	ident = v.GoVisitor.VisitIdentifier(ident, p).(*java.Identifier)
	if ident.Name == "foo" {
		ident = ident.WithName("bar")
	}
	return ident
}

func TestRecipeRename(t *testing.T) {
	spec := test.NewRecipeSpec().WithRecipe(&renameFooToBar{})
	spec.RewriteRun(t,
		test.Golang(`
			package main

			func foo() {
			}
		`, `
			package main

			func bar() {
			}
		`),
	)
}

func TestRecipeNoChange(t *testing.T) {
	spec := test.NewRecipeSpec().WithRecipe(&renameFooToBar{})
	spec.RewriteRun(t,
		test.Golang(`
			package main

			func hello() {
			}
		`),
	)
}

// TestSearchRecipeViaRewriteRun verifies that RewriteRun prints SearchResult
// markers as /*~~(...)~~>*/ comments by default, so search-style recipes can
// be tested with the same convention used in Java and TypeScript.
func TestSearchRecipeViaRewriteRun(t *testing.T) {
	spec := test.NewRecipeSpec().WithRecipe(&findFoo{})
	spec.RewriteRun(t,
		test.GolangRaw(
			"package main\n\nfunc foo() {\n}\n",
			"package main\n\nfunc /*~~(found foo)~~>*/foo() {\n}\n",
		),
	)
}

// TestSearchRecipeWithSanitizedMarkerPrinter verifies that
// WithMarkerPrinter(SanitizedMarkerPrinter) lets tests opt out of marker
// rendering — matching Java's spec.markerPrinter(SANITIZED).
func TestSearchRecipeWithSanitizedMarkerPrinter(t *testing.T) {
	spec := test.NewRecipeSpec().
		WithRecipe(&findFoo{}).
		WithMarkerPrinter(printer.SanitizedMarkerPrinter)
	spec.RewriteRun(t,
		test.GolangRaw(
			"package main\n\nfunc foo() {\n}\n",
			"package main\n\nfunc foo() {\n}\n",
		),
	)
}

func TestSearchRecipeWithMarkerPrinting(t *testing.T) {
	r := &findFoo{}
	editor := r.Editor()

	src := "package main\n\nfunc foo() {\n}\n"
	p := parser.NewGoParser()
	cu, err := p.Parse("test.go", src)
	require.NoError(t, err)

	ctx := recipe.NewExecutionContext()
	result := editor.Visit(cu, ctx)

	// Print with default marker printer — should show search result comment
	output := printer.PrintWithMarkers(result, printer.DefaultMarkerPrinter)
	expected := "package main\n\nfunc /*~~(found foo)~~>*/foo() {\n}\n"
	assert.Equal(t, expected, output, "marker output mismatch")

	// Print without markers — should be original source
	plain := printer.Print(result)
	assert.Equal(t, plain, src, "plain print should match original source")

	// Print with sanitized printer — should strip markers
	sanitized := printer.PrintWithMarkers(result, printer.SanitizedMarkerPrinter)
	assert.Equal(t, sanitized, src, "sanitized print should match original source")
}

func TestRecipeDescriptor(t *testing.T) {
	r := &findFoo{}
	desc := recipe.Describe(r)

	assert.Equalf(t, "org.openrewrite.golang.test.FindFoo", desc.Name, "expected name %q", "org.openrewrite.golang.test.FindFoo")
	assert.Equalf(t, "Find foo identifiers", desc.DisplayName, "expected displayName %q", "Find foo identifiers")
	if desc.EstimatedEffortPerOccurrence != 5*time.Minute {
		t.Errorf("expected 5 minute default effort, got %v", desc.EstimatedEffortPerOccurrence)
	}
}

func TestRegistryActivate(t *testing.T) {
	reg := recipe.NewRegistry()

	// Each module provides an Activate function
	activateSearch := func(r *recipe.Registry) {
		golang := recipe.CategoryDescriptor{DisplayName: "Go"}
		search := recipe.CategoryDescriptor{DisplayName: "Search"}
		r.Register(&findFoo{}, golang, search)
	}
	activateRefactoring := func(r *recipe.Registry) {
		golang := recipe.CategoryDescriptor{DisplayName: "Go"}
		r.Register(&renameFooToBar{}, golang)
	}

	reg.Activate(activateSearch, activateRefactoring)

	found, ok := reg.FindRecipe("org.openrewrite.golang.test.FindFoo")
	require.True(t, ok, "expected to find FindFoo recipe")
	assert.Equalf(t, "Find foo identifiers", found.Descriptor.DisplayName, "expected displayName %q", "Find foo identifiers")

	// All recipes
	all := reg.AllRecipes()
	assert.Len(t, all, 2, "expected 2 recipes")

	// Categories
	cats := reg.Categories()
	require.Len(t, cats, 1, "expected 1 top-level category")
	assert.Equalf(t, "Go", cats[0].DisplayName, "expected top-level category 'Go', got %q", cats[0].DisplayName)
	assert.Lenf(t, cats[0].Recipes, 1, "expected 1 recipe directly in Go category, got %d", len(cats[0].Recipes))
	require.Lenf(t, cats[0].Subcategories, 1, "expected 1 subcategory in Go, got %d", len(cats[0].Subcategories))
	assert.Equalf(t, "Search", cats[0].Subcategories[0].DisplayName, "expected subcategory 'Search', got %q", cats[0].Subcategories[0].DisplayName)
}

func TestRegistryReflectConstructor(t *testing.T) {
	reg := recipe.NewRegistry()

	reg.Activate(func(r *recipe.Registry) {
		r.Register(&renameFooToBar{}, recipe.CategoryDescriptor{DisplayName: "Go"})
	})

	found, ok := reg.FindRecipe("org.openrewrite.golang.test.RenameFooToBar")
	require.True(t, ok, "expected to find recipe")

	// Constructor auto-derived from prototype via reflection
	instance := found.Constructor(nil)
	assert.Equal(t, "org.openrewrite.golang.test.RenameFooToBar", instance.Name(), "unexpected name")
}

func TestFencedMarkerPrinting(t *testing.T) {
	r := &findFoo{}
	editor := r.Editor()

	src := "package main\n\nfunc foo() {\n}\n"
	p := parser.NewGoParser()
	cu, err := p.Parse("test.go", src)
	require.NoError(t, err)

	ctx := recipe.NewExecutionContext()
	result := editor.Visit(cu, ctx)

	// Print with fenced printer — should show {{uuid}} delimiters
	output := printer.PrintWithMarkers(result, printer.FencedMarkerPrinter)
	assert.NotEqual(t, output, src, "expected fenced markers in output, but output is unchanged")
	// Fenced output should contain UUID-style markers
	if len(output) <= len(src) {
		t.Error("fenced output should be longer than original source")
	}
}

File: rewrite-csharp/csharp/OpenRewrite.Tests/Recipe/RecipeTest.cs
/*
 * Copyright 2026 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using OpenRewrite.Core;
using OpenRewrite.CSharp;
using OpenRewrite.Java;
using OpenRewrite.Test;

namespace OpenRewrite.Tests.Recipe;

public class RecipeTest : RewriteTest
{
    [Fact]
    public void RenameClass()
    {
        RewriteRun(
            spec => spec.SetRecipe(new RenameClassRecipe { From = "Foo", To = "Bar" }),
            CSharp(
                "class Foo { }",
                "class Bar { }"
            )
        );
    }

    [Fact]
    public void NoChangeWhenClassNotFound()
    {
        RewriteRun(
            spec => spec.SetRecipe(new RenameClassRecipe { From = "Foo", To = "Bar" }),
            CSharp("class Baz { }")
        );
    }

    [Fact]
    public void RenameClassInNamespace()
    {
        RewriteRun(
            spec => spec.SetRecipe(new RenameClassRecipe { From = "Foo", To = "Bar" }),
            CSharp(
                """
                namespace MyApp
                {
                    class Foo { }
                }
                """,
                """
                namespace MyApp
                {
                    class Bar { }
                }
                """
            )
        );
    }

    [Fact]
    public void NoOpVisitorPreservesReferenceEquality()
    {
        var parser = new CSharpParser();
        var source = parser.Parse("class C { void M() { var x = 1; } }");
        var visitor = new CSharpVisitor<OpenRewrite.Core.ExecutionContext>();
        var result = visitor.Visit(source, new OpenRewrite.Core.ExecutionContext());
        Assert.True(ReferenceEquals(source, result),
            $"No-op visitor should preserve reference equality. " +
            $"source type={source.GetType().Name}, result type={result?.GetType().Name}");
    }

    [Fact]
    public void RecipeDescriptorHasOptions()
    {
        var recipe = new RenameClassRecipe { From = "Foo", To = "Bar" };
        var descriptor = recipe.GetDescriptor();

        Assert.Equal("Rename class", descriptor.DisplayName);
        Assert.Equal(2, descriptor.Options.Count);
        Assert.Equal("From", descriptor.Options[0].Name);
        Assert.Equal("String", descriptor.Options[0].Type);
        Assert.Equal("The class name to rename from.", descriptor.Options[0].Description);
        Assert.Equal("Foo", descriptor.Options[0].Example);
        Assert.True(descriptor.Options[0].Required);
        Assert.False(descriptor.Options[0].Secret);
        Assert.Equal("Foo", descriptor.Options[0].Value);
        Assert.Equal("To", descriptor.Options[1].Name);
        Assert.Equal("Bar", descriptor.Options[1].Value);
    }

    [Fact]
    public void SecretOptionPropagatesAndCanBeRedacted()
    {
        var recipe = new RecipeWithSecret { ApiToken = "hunter2" };
        var descriptor = recipe.GetDescriptor();

        Assert.Single(descriptor.Options);
        var opt = descriptor.Options[0];
        Assert.Equal("ApiToken", opt.Name);
        Assert.True(opt.Secret);
        // Raw value is preserved on the source-level descriptor so RPC + recipe execution
        // still see it. Persistence boundaries must call WithRedactedSecretValue().
        Assert.Equal("hunter2", opt.Value);

        var redacted = opt.WithRedactedSecretValue();
        Assert.True(redacted.Secret);
        Assert.Null(redacted.Value);
        Assert.Equal("ApiToken", redacted.Name);
        Assert.Equal("API token", redacted.DisplayName);
    }

    [Fact]
    public void WithRedactedSecretValueIsNoOpForNonSecret()
    {
        var recipe = new RenameClassRecipe { From = "Foo", To = "Bar" };
        var opt = recipe.GetDescriptor().Options[0];
        Assert.Same(opt, opt.WithRedactedSecretValue());
    }

}

class RecipeWithSecret : OpenRewrite.Core.Recipe
{
    [Option(DisplayName = "API token", Description = "API token used by the recipe.", Secret = true)]
    public required string ApiToken { get; init; }

    public override string DisplayName => "Recipe with secret";
    public override string Description => "Recipe with secret.";

    public override JavaVisitor<OpenRewrite.Core.ExecutionContext> GetVisitor() =>
        new CSharpVisitor<OpenRewrite.Core.ExecutionContext>();
}

class RenameClassRecipe : OpenRewrite.Core.Recipe
{
    [Option(DisplayName = "From", Description = "The class name to rename from.", Example = "Foo")]
    public required string From { get; init; }

    [Option(DisplayName = "To", Description = "The class name to rename to.", Example = "Bar")]
    public required string To { get; init; }

    public override string DisplayName => "Rename class";
    public override string Description => $"Renames class `{From}` to `{To}`.";

    public override JavaVisitor<OpenRewrite.Core.ExecutionContext> GetVisitor() => new RenameClassVisitor(From, To);

    private class RenameClassVisitor : CSharpVisitor<OpenRewrite.Core.ExecutionContext>
    {
        private readonly string _from;
        private readonly string _to;

        public RenameClassVisitor(string from, string to)
        {
            _from = from;
            _to = to;
        }

        public override J VisitClassDeclaration(ClassDeclaration cd, OpenRewrite.Core.ExecutionContext ctx)
        {
            cd = (ClassDeclaration)base.VisitClassDeclaration(cd, ctx);
            if (cd.Name.SimpleName == _from)
            {
                return cd.WithName(cd.Name.WithSimpleName(_to));
            }

            return cd;
        }
    }
}
